### ISSUE TITLE
[go_router] Push doesn't update the URL

### ISSUE THREAD

**Issue by Gieted** [flutter/flutter#134318]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

In Flutter Web:

1.  Do something like `context.push('/settings');`
2. See url doesn't update as page navigates

### Expected results

The url should change to 'https://mydomain.com/settings'

### Actual results

The url doesn't change at all.

This issue is actually exactly the same as #115962

### Code sample

`context.push('/settings');`

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.13.3, on Microsoft Windows [Version 10.0.22621.2134], locale pl-PL)
    • Flutter version 3.13.3 on channel stable at C:\tools\flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 2524052335 (2 days ago), 2023-09-06 14:32:31 -0700
    • Engine revision b8d35810e9
    • Dart version 3.1.1
    • DevTools version 2.25.0

[✓] Windows Version (Installed version of Windows is version 10 or higher)

[✗] Android toolchain - develop for Android devices
    ✗ Unable to locate Android SDK.
      Install Android Studio from: https://developer.android.com/studio/index.html
      On first launch it will assist you in installing the Android SDK components.
      (or visit https://flutter.dev/docs/get-started/install/windows#android-setup for detailed instructions).
      If the Android SDK has been installed to a custom location, please use
      `flutter config --android-sdk` to update to that location.


[✓] Chrome - develop for the web
    • CHROME_EXECUTABLE = C:\Program Files\Chromium\Application\chrome.exe

[✓] Visual Studio - develop Windows apps (Visual Studio Community 2022 17.4.3)
    • Visual Studio at C:\Program Files\Microsoft Visual Studio\2022\Community
    • Visual Studio Community 2022 version 17.4.33205.214
    • Windows 10 SDK version 10.0.22000.0

[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.dev/docs/get-started/install/windows#android-setup for detailed instructions).

[✓] IntelliJ IDEA Ultimate Edition (version 2023.2)
    • IntelliJ at C:\Users\Gieted\AppData\Local\Programs\IntelliJ IDEA Ultimate
    • Flutter plugin version 75.1.4
    • Dart plugin version 232.9559.10

[✓] Connected device (3 available)
    • Windows (desktop) • windows • windows-x64    • Microsoft Windows [Version 10.0.22621.2134]
    • Chrome (web)      • chrome  • web-javascript • unknown
    • Edge (web)        • edge    • web-javascript • Microsoft Edge 116.0.1938.62

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 2 categories.
```

</details>

**Comment by Gieted** [comment]
Ok, I now saw this [comment](https://github.com/flutter/flutter/issues/99112#issuecomment-1563213451).

> We can have a flag in GoRouter constructor to restore this behavior if someone really want the url to change, but we will document against using this flag.

Has this flag already been added?

And also, I don't understand why the "declarative style" of navigation has been chosen as a default for this package. Native Android and Web use stack-based navigation, because it's far more natural for the user. I also don't understand why you consider imperative routes as "not deep-linkable".

**Comment by darshankawar** [comment]
@Gieted 
Can you take a look at [this](https://docs.google.com/document/d/1vjupshmFJtfGSOppZxp7Tzkq7dotcLxCcpdluuNYe1o/edit?pli=1&resourcekey=0-aS66t4OcDTjJW50s-veSzQ) migration guide and check if it helps in your case or not ?

**Comment by Gieted** [comment]
> @Gieted Can you take a look at [this](https://docs.google.com/document/d/1vjupshmFJtfGSOppZxp7Tzkq7dotcLxCcpdluuNYe1o/edit?pli=1&resourcekey=0-aS66t4OcDTjJW50s-veSzQ) migration guide and check if it helps in your case or not ?

No, it's actually completely unrelated to my problem.
I've ended up writing my own router delegate and would highly recommend doing so to anyone creating an app with complex navigation. It's not as hard as it initially looks and gives you the total control over rooting in your application.

However, I would still want to ask go_router maintainers how they think named imperative routes should be handled. I can completely understand, that on native mobile on updating the url is acceptable, since you don't see the url bar anyway. However, on web things are a little more complicated. For e.g. a page can be refreshed, and the user will feel like the action of refreshing the page navigates them away from where currently they were, because all unnamed routes pushed to the stack are lost. Additionally, you cannot save such page as a bookmark or send it as a link.

Basically, only the most obscure web pages do not use different url paths to represent different parts of the application. I think that a package like go_router should definitely have some way to update url, without losing the back-stack.

**Comment by darshankawar** [comment]
Thanks for the update and feedback. Keeping it open based on the report.

**Comment by chunhtai** [comment]
The decision is made after 8 and up. You can set [optionURLReflectsImperativeAPIsto](https://pub.dev/documentation/go_router/latest/go_router/GoRouter/optionURLReflectsImperativeAPIs.html) true if you want to have the old behavior. 
See flutter.dev/go/go-router-v8-breaking-changes
